internal/rbd/group_controller.go

```diff
@@ -0,0 +1,214 @@
+			log.DebugLog(ctx, "removing temporary volume group %q", vg)
+
+			vgErr := vg.Delete(ctx)
+			if vgErr != nil {
+				log.ErrorLog(ctx, "failed to remove temporary volume group %q: %v", vg, vgErr)
+			}
+
+			vg.Destroy(ctx)
+		}
+	}()
+	for i, id := range req.GetSourceVolumeIds() {
+		var vol types.Volume
+		vol, err = mgr.GetVolumeByID(ctx, id)
+		if err != nil {
+			return nil, status.Errorf(
+				codes.InvalidArgument,
+				"failed to find required volume %q for volume group snapshot %q: %s",
+				id,
+				vgsName,
+				err.Error())
+		}
+		volumes[i] = vol
+	}
+
+	log.DebugLog(ctx, "all %d Volumes for VolumeGroup %q have been found", len(volumes), vgsName)
+
+	groupSnapshot, err = mgr.GetVolumeGroupSnapshotByName(ctx, vgsName)
+	if groupSnapshot != nil {
+		csiVGS, csiErr := groupSnapshot.ToCSI(ctx)
+		if csiErr != nil {
+			return nil, status.Error(codes.Aborted, csiErr.Error())
+		}
+
+		return &csi.CreateVolumeGroupSnapshotResponse{
+			GroupSnapshot: csiVGS,
+		}, nil
+	}
+	if err != nil {
+		log.DebugLog(ctx, "need to create new volume group snapshot, "+
+			"failed to get existing one with name %q: %v", err)
+	}
+
+	// create a temporary VolumeGroup with a different name
+	vg, err = mgr.CreateVolumeGroup(ctx, vgName)
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"failed to create volume group %q: %s",
+			vgName,
+			err.Error())
+	}
+
+	log.DebugLog(ctx, "VolumeGroup %q has been created: %+v", vgName, vg)
+
+	// add images to the group
+	for _, volume := range volumes {
+		err = vg.AddVolume(ctx, volume)
+		if err != nil {
+			return nil, status.Error(codes.Aborted, err.Error())
+		}
+	}
+
+	groupSnapshot, err = mgr.CreateVolumeGroupSnapshot(ctx, vg, vgsName)
+	if err != nil {
+		return nil, status.Error(codes.Aborted, err.Error())
+	}
+	defer groupSnapshot.Destroy(ctx)
+
+	csiVGS, err := groupSnapshot.ToCSI(ctx)
+	if err != nil {
+		return nil, status.Error(codes.Aborted, err.Error())
+	}
+
+	return &csi.CreateVolumeGroupSnapshotResponse{
+		GroupSnapshot: csiVGS,
+	}, nil
+}
+
+func (cs *ControllerServer) DeleteVolumeGroupSnapshot(
+	ctx context.Context,
+	req *csi.DeleteVolumeGroupSnapshotRequest,
+) (*csi.DeleteVolumeGroupSnapshotResponse, error) {
+	// FIXME: more checking of the request in needed
+	// 1. verify that all snapshots in the request are all snapshots in the group
+	// 2. delete the group
+
+	mgr := NewManager(cs.Driver.GetInstanceID(), nil, req.GetSecrets())
+	defer mgr.Destroy(ctx)
+
+	groupSnapshot, err := mgr.GetVolumeGroupSnapshotByID(ctx, req.GetGroupSnapshotId())
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"failed to get volume group snapshot with id %q: %v",
+			req.GetGroupSnapshotId(), err)
+	}
+	defer groupSnapshot.Destroy(ctx)
+
+	err = groupSnapshot.Delete(ctx)
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"failed to delete volume group snapshot %q: %v",
+			groupSnapshot, err)
+	}
```

Comment:
This is going to take care of deleting all the clones isn't it? just checking to make sure we are good here

Reply:
Yes, that is a requirement from the CSI specification.